jetty-server/src/main/java/org/eclipse/jetty/server/BlockingContentProducer.java

```diff
@@ -0,0 +1,133 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.concurrent.Semaphore;
+
+class BlockingContentProducer implements ContentProducer
+{
+    private final Semaphore _semaphore = new Semaphore(0);
+    private final AsyncContentProducer _asyncContentProducer;
+    private final HttpChannel _httpChannel;
+
+    public BlockingContentProducer(HttpChannel httpChannel, AsyncContentProducer delegate)
+    {
+        _asyncContentProducer = delegate;
+        _httpChannel = httpChannel;
+    }
+
+    @Override
+    public void recycle()
+    {
+        _asyncContentProducer.recycle();
+    }
+
+    @Override
+    public int available()
+    {
+        return _asyncContentProducer.available();
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        return _asyncContentProducer.hasContent();
+    }
+
+    @Override
+    public boolean isError()
+    {
+        return _asyncContentProducer.isError();
+    }
+
+    @Override
+    public void checkMinDataRate()
+    {
+        _asyncContentProducer.checkMinDataRate();
+    }
+
+    @Override
+    public long getRawContentArrived()
+    {
+        return _asyncContentProducer.getRawContentArrived();
+    }
+
+    @Override
+    public boolean consumeAll(Throwable x)
+    {
+        return _asyncContentProducer.consumeAll(x);
+    }
+
+    @Override
+    public HttpInput.Content nextContent()
+    {
+        while (true)
+        {
+            HttpInput.Content content = _asyncContentProducer.nextContent();
+            if (content != null)
+                return content;
+
+            _semaphore.drainPermits();
+            _httpChannel.getState().onReadUnready();
```

Comment:
Why is not this call needed for the async case?
Feels that blocking may be implemented in terms of async with just the semaphore, but only in this case we need an additional state change?

Reply:
I find it strange that the blocking impl is half wrapping the async impl and half calling HttpChannelState directly.
The blocking impl should be 100% based on the async impl and any state changes that result should be done by the async impl

Reply:
I don't think `drainPermits()` is needed, and I also think that
```java
_httpChannel.getState().onReadUnready();
if (_httpChannel.needContent())
  continue;
```
can be replaced with
```java
if (_asyncContentProducer.isReady())
  continue;
```

I'll try both changes.

Reply:
Done.